File: python/tests/test_batched_shape_rule.py
"""Tests for the shape rule that selects batched-raw mode.

Python dispatches on ``ndim``: a 2-D array is a batch whatever its
second dimension, so ``(M, 1)`` is M rows of one element each. MATLAB
dispatches on a matrix with both dimensions greater than one, so an
M-by-1 column is a vector there and is broadcast as one shared
multiset. The two rules are each idiomatic in their own language and
are documented as a deliberate divergence in the entry-point
docstrings; these tests pin the Python side of it, and pin the
NaN-padded spelling that reads the same way in both languages. The
MATLAB twin is ``tests/test_batched_shape_rule.m``.
"""

import numpy as np
import pytest

import mpt


P = 1200.0
SIGMA = 10.0
SCALE = np.array([0.0, 200.0, 400.0, 500.0, 700.0, 900.0, 1100.0])
PROBES = np.array([0.0, 100.0, 200.0, 300.0])


def _context(n):
    return np.tile(SCALE, (n, 1))


class TestColumnVectorIsABatch:

    def test_column_of_probes_gives_one_result_per_row(self):
        n = PROBES.size
        s = mpt.sim_maet(_context(n), None, PROBES.reshape(n, 1), None,
                         SIGMA, 1, False, True, P, verbose=False)
        assert np.asarray(s).shape == (n,)

    def test_the_results_are_not_all_equal(self):
        """The diagnostic that caught the divergence: broadcasting the
        column as one shared multiset would make the profile flat."""
        n = PROBES.size
        s = np.asarray(mpt.sim_maet(_context(n), None,
                                    PROBES.reshape(n, 1), None,
                                    SIGMA, 1, False, True, P,
                                    verbose=False))
        assert np.ptp(s) > 1e-3

    def test_row_vector_broadcasts_instead(self):
        """A 1-D operand, or a ``(1, K)`` one, is the broadcast form."""
        n = 3
        one = np.array([0.0, 400.0, 700.0])
        flat = mpt.sim_maet(_context(n), None, one, None,
                            SIGMA, 1, False, True, P, verbose=False)
        two_d = mpt.sim_maet(_context(n), None, one.reshape(1, 3), None,
                             SIGMA, 1, False, True, P, verbose=False)
        assert np.allclose(np.asarray(flat), np.asarray(two_d))
        assert np.asarray(flat).shape == (n,)


class TestNanPaddedSpelling:

    def test_padding_agrees_with_the_bare_column(self):
        """The spelling that carries across to MATLAB: pad to two
        columns with NaN. The padding is stripped per row, so the
        values are those of the one-element rows."""
        n = PROBES.size
        bare = mpt.sim_maet(_context(n), None, PROBES.reshape(n, 1), None,
                            SIGMA, 1, False, True, P, verbose=False)
        padded = mpt.sim_maet(
            _context(n), None,
            np.column_stack([PROBES, np.full(n, np.nan)]), None,
            SIGMA, 1, False, True, P, verbose=False)
        assert np.allclose(np.asarray(bare), np.asarray(padded))

    def test_padded_rows_match_probe_by_probe_calls(self):
        n = PROBES.size
        padded = np.asarray(mpt.sim_maet(
            _context(n), None,
            np.column_stack([PROBES, np.full(n, np.nan)]), None,
            SIGMA, 1, False, True, P, verbose=False))
        one_by_one = np.array([
            mpt.sim_maet(SCALE, None, np.array([x]), None,
                         SIGMA, 1, False, True, P, verbose=False)
            for x in PROBES])
        assert np.allclose(padded, one_by_one)

    def test_padding_survives_the_spectrum_keyword(self):
        n = PROBES.size
        spectrum = ['harmonic', 8, 'powerlaw', 1]
        bare = mpt.sim_maet(_context(n), None, PROBES.reshape(n, 1), None,
                            SIGMA, 1, False, True, P,
                            spectrum=spectrum, verbose=False)
        padded = mpt.sim_maet(
            _context(n), None,
            np.column_stack([PROBES, np.full(n, np.nan)]), None,
            SIGMA, 1, False, True, P,
            spectrum=spectrum, verbose=False)
        assert np.allclose(np.asarray(bare), np.asarray(padded))


class TestEvalAndEntropy:

    def test_eval_reads_a_column_as_rows(self):
        vals = mpt.eval_maet(PROBES.reshape(-1, 1), None, SIGMA, 1,
                             False, True, P, np.array([[0.0]]),
                             verbose=False)
        assert np.asarray(vals).shape == (PROBES.size, 1)

    def test_entropy_reads_a_column_as_rows(self):
        H = mpt.entropy_maet(PROBES.reshape(-1, 1), None, SIGMA, 1,
                             False, True, P, n_points_per_dim=120,
                             verbose=False)
        assert np.asarray(H).shape == (PROBES.size,)

File: matlab/tests/test_batched_shape_rule.m
%% test_batched_shape_rule.m — the shape rule that selects batched-raw mode
%
%  MATLAB dispatches on a matrix with both dimensions greater than one,
%  so an M-by-1 column is a vector and is broadcast as one multiset
%  shared by every row. Python dispatches on ndim, where an (M, 1) array
%  is M rows of one element each. The two rules are each idiomatic in
%  their own language and are documented as a deliberate divergence in
%  the entry-point docstrings; these tests pin the MATLAB side of it,
%  and pin the NaN-padded spelling that reads the same way in both
%  languages.
%
%  Mirror of Python tests/test_batched_shape_rule.py.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    addpath(fileparts(fileparts(mfilename('fullpath'))));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

PERIOD = 1200;
SIGMA  = 10;
SCALE  = [0 200 400 500 700 900 1100];
PROBES = [0 100 200 300];

nRows   = numel(PROBES);
context = repmat(SCALE, nRows, 1);

% --- a column is a vector, not a batch ------------------------------------

sCol = simMaet(context, [], PROBES(:), [], SIGMA, 1, false, true, ...
               PERIOD, 'verbose', false);
results{end+1,1} = ['simMaet: a column operand is broadcast as one ' ...
                    'shared multiset'];
results{end,2}   = numel(sCol) == nRows ...
                   && (max(sCol(:)) - min(sCol(:))) < 1e-12;

sRow = simMaet(context, [], PROBES, [], SIGMA, 1, false, true, ...
               PERIOD, 'verbose', false);
results{end+1,1} = 'simMaet: row and column operands agree, both broadcasting';
results{end,2}   = max(abs(sCol(:) - sRow(:))) < 1e-12;

% --- the NaN-padded spelling ----------------------------------------------

probes = [PROBES(:), nan(nRows, 1)];
sPad = simMaet(context, [], probes, [], SIGMA, 1, false, true, ...
               PERIOD, 'verbose', false);
results{end+1,1} = 'simMaet: NaN padding gives one result per probe row';
results{end,2}   = numel(sPad) == nRows ...
                   && (max(sPad(:)) - min(sPad(:))) > 1e-3;

% The padding is stripped per row, so each row is the one-tone probe it
% was written to be: the profile is the same as comparing each probe
% separately.
sOneByOne = zeros(nRows, 1);
for i = 1:nRows
    sOneByOne(i) = simMaet(SCALE, [], PROBES(i), [], SIGMA, 1, ...
                           false, true, PERIOD, 'verbose', false);
end
results{end+1,1} = 'simMaet: the padded rows match probe-by-probe calls';
results{end,2}   = max(abs(sPad(:) - sOneByOne)) < 1e-10;

% The same holds with a spectrum applied inside the call, which is how
% the probe-tone demo uses it.
SPECTRUM = {'harmonic', 8, 'powerlaw', 1};
sPadSpec = simMaet(context, [], probes, [], SIGMA, 1, false, true, ...
                   PERIOD, 'spectrum', SPECTRUM, 'verbose', false);
% The per-probe reference is itself a batched call, of two identical
% rows. 'spectrum' is a batched-raw name-value in MATLAB -- on a scalar
% operand the caller is told to apply addSpectra beforehand -- so a
% one-probe-at-a-time loop cannot express the comparison here, and the
% smallest batch that can is a pair.
sSpecOneByOne = zeros(nRows, 1);
for i = 1:nRows
    sPair = simMaet(repmat(SCALE, 2, 1), [], ...
                    [PROBES(i) NaN; PROBES(i) NaN], [], SIGMA, 1, ...
                    false, true, PERIOD, ...
                    'spectrum', SPECTRUM, 'verbose', false);
    sSpecOneByOne(i) = sPair(1);
end
results{end+1,1} = 'simMaet: NaN padding survives the spectrum name-value';
results{end,2}   = max(abs(sPadSpec(:) - sSpecOneByOne)) < 1e-10;

% --- the same rule in evalMaet and entropyMaet ----------------------------

X = [0 100 200];
valsCol = evalMaet(PROBES(:), [], SIGMA, 1, false, true, PERIOD, X, ...
                   'verbose', false);
results{end+1,1} = ['evalMaet: a column takes the single multiset raw ' ...
                    'path'];
results{end,2}   = isequal(size(valsCol), [1 numel(X)]);

valsPad = evalMaet([PROBES(:), nan(nRows, 1)], [], SIGMA, 1, false, ...
                   true, PERIOD, X, 'verbose', false);
results{end+1,1} = 'evalMaet: NaN padding gives one row per probe';
results{end,2}   = isequal(size(valsPad), [nRows numel(X)]);

HCol = entropyMaet(PROBES(:), [], SIGMA, 1, false, true, PERIOD, ...
                   'nPointsPerDim', 120, 'verbose', false);
results{end+1,1} = 'entropyMaet: a column takes the single multiset raw path';
results{end,2}   = isscalar(HCol);

HPad = entropyMaet([PROBES(:), nan(nRows, 1)], [], SIGMA, 1, false, ...
                   true, PERIOD, 'nPointsPerDim', 120, 'verbose', false);
results{end+1,1} = 'entropyMaet: NaN padding gives one entropy per probe';
results{end,2}   = numel(HPad) == nRows;


if standalone
    nPass = 0; nFail = 0;
    for i = 1:size(results, 1)
        if results{i,2}
            nPass = nPass + 1;
            fprintf('  PASS  %s\n', results{i,1});
        else
            nFail = nFail + 1;
            fprintf('  FAIL  %s\n', results{i,1});
        end
    end
    fprintf(['\n=== test_batched_shape_rule: %d passed, %d failed ' ...
             '(of %d) ===\n\n'], nPass, nFail, nPass + nFail);
    clear cleanupDefaults
    if nFail > 0
        error('test_batched_shape_rule:failed', '%d test(s) failed.', nFail);
    end
end
